Update PR with review comments

malhotra5 opened the pull request:
# Summary
This PR addresses #27 

- Updated `openhands_resolver/resolve_issues.py` with flag `--issue-type pr`
- Recreates PR locally based on PR identity (number and repo name)
- Downloads contextual information
    - original linked issues the PR would close
    - review comments
- merges comment threads into singular string
- all unresolved comments are passed (together) as the problem statement into custom prompt `openhands_resolver/prompts/resolve/basic-followup.jinja`
- same workflow as before is maintained for codebase modifications
- output.jsonl indicates if all comments were resolved successfully; success indicator were not implemented on a per comment basis

Note: version bumping was not performed on this PR as described in `.openhands_instructions`

# Future work to be done
1. Currently all the unresolved comments are passed at once for a given PR. Possibly process unresolved comment separately? This may lead to following issues - 
    a.  Should every resolved comment open a new PR request, or update the current one?
    b. What happens when 3/5 comments are resolved? How can the user specify which code changes to push on a per comment basis?
    c. How are comment dependencies handled? For instance, if two comments address the same part of the codebase, handling them separately may create merge conflicts. Do we expect the user to perform incremental commenting to mitigate this?

neubig commented on openhands_resolver/github_issue.py:
```diff
@@ -1,4 +1,5 @@
 from pydantic import BaseModel
+from typing import Optional
```
```suggestion
```
We can do `XXX | None` starting in Python 3.10, so let's use this. It's the recommended way now: https://stackoverflow.com/a/69440627

neubig commented on openhands_resolver/github_issue.py:
```diff
@@ -7,3 +8,6 @@ class GithubIssue(BaseModel):
     number: int
     title: str
     body: str
+    closing_issues: Optional[list[str]] = []
+    review_comments: Optional[list[str]] = []
+    head_branch: Optional[str] = None
```
```suggestion
    closing_issues: list[str] | None = None
    review_comments: list[str] | None = None
    head_branch: str | None = None
```
I think `None` is OK here as the default to indicate that it hasn't been initialized (to distinguish with the case where it has been initialized).

neubig commented on openhands_resolver/prompts/resolve/basic-followup.jinja:
```diff
@@ -0,0 +1,17 @@
+The current code is an attempt at fixing a set of issues. The code is not satisfactory and follow up feedback have been provided to address this.
```
```suggestion
The current code is an attempt at fixing one or more issues. The code is not satisfactory and follow up feedback have been provided to address this.
```

neubig commented on openhands_resolver/prompts/resolve/basic-followup.jinja:
```diff
@@ -0,0 +1,17 @@
+Please update the code based on the feedback for the repository in /workspace.
+An environment has been set up for you to start working. You may assume all necessary tools are installed.
+
+Issues addressed 
```
```suggestion
# Issues addressed 
```

neubig commented on openhands_resolver/resolve_issues.py:
```diff
@@ -656,6 +812,14 @@ def main():
+
+    parser.add_argument(
+        "--issue-type",
+        type=str,
+        default=None,
```
```suggestion
        type=str,
        default="issue",
        choices=["issue", "pr"],
```

neubig commented on openhands_resolver/resolve_issues.py:
```diff
@@ -197,9 +198,16 @@ def get_instruction(
     issue: GithubIssue,
     prompt_template: str,
     repo_instruction: str | None = None,
+    issue_type: str | None = None
```
```suggestion
    issue_type: str,
    repo_instruction: str | None = None,
```
We probably don't need `None` issue types, see below.

neubig commented on openhands_resolver/resolve_issues.py:
```diff
@@ -197,9 +198,16 @@ def get_instruction(
 ):  # Prepare instruction
     template = jinja2.Template(prompt_template)
-    instruction = template.render(body=issue.body, repo_instruction=repo_instruction)
+    if issue_type == None:
+        instruction = template.render(body=issue.body, repo_instruction=repo_instruction)
+    else:
+        issues_context = json.dumps(issue.closing_issues, indent=4)
+        comment_chain = json.dumps(issue.review_comments, indent=4)
+        instruction = template.render(issues=issues_context, body=comment_chain, repo_instruction=repo_instruction)
```
```suggestion
    if issue_type == "issue":
        instruction = template.render(body=issue.body, repo_instruction=repo_instruction)
    elif issue_type == "pr":
        if issue.closing_issues is None or issue.review_comments is None:
            raise ValueError(f"issue.closing_issues or issue.review_comments is None")
        issues_context = json.dumps(issue.closing_issues, indent=4)
        comment_chain = json.dumps(issue.review_comments, indent=4)
        instruction = template.render(issues=issues_context, body=comment_chain, repo_instruction=repo_instruction)
    else:
        raise ValueError(f"Invalid issue type {issue_type}")
```

neubig commented on openhands_resolver/resolve_issues.py:
```diff
@@ -254,6 +262,7 @@ async def process_issue(
     prompt_template: str,
     repo_instruction: str | None = None,
     reset_logger: bool = True,
+    issue_type: str | None = None
```
```suggestion
    issue_type: str,
    repo_instruction: str | None = None,
    reset_logger: bool = True,
```

neubig commented on openhands_resolver/resolve_issues.py:
```diff
@@ -263,7 +272,11 @@ async def process_issue(
         logger.info(f'Starting fixing issue {issue.number}.')
 
-    workspace_base = os.path.join(output_dir, "workspace", f"issue_{issue.number}")
+    if issue_type == None:
+        workspace_base = os.path.join(output_dir, "workspace", f"issue_{issue.number}")
+    else:
+        workspace_base = os.path.join(output_dir, "workspace", f"issue_{issue_type}_{issue.number}")
```
```suggestion
    workspace_base = os.path.join(output_dir, "workspace", f"{issue_type}_{issue.number}")
```

neubig commented on openhands_resolver/resolve_issues.py:
```diff
@@ -292,7 +305,7 @@ async def process_issue(
     initialize_runtime(runtime)
 
-    instruction = get_instruction(issue, prompt_template, repo_instruction)
+    instruction = get_instruction(issue, prompt_template, repo_instruction, issue_type)
```
```suggestion
    instruction = get_instruction(issue, prompt_template, issue_type, repo_instruction)
```

neubig commented on tests/test_resolve_issues.py:
```diff
@@ -120,12 +126,96 @@ def test_download_issues_from_github():
     assert [issue.title for issue in issues] == ["Issue 1", "Issue 2"]
+    assert [len(issue.review_comments) for issue in issues] == [0, 0]
+    assert [len(issue.closing_issues) for issue in issues] == [0, 0]
+
```
could we split this into a separate test?

neubig commented on openhands_resolver/resolve_issues.py:
```diff
@@ -697,6 +854,23 @@ def main():
     if my_args.issue_numbers:
         issue_numbers = [int(number) for number in my_args.issue_numbers.split(",")]
 
+    issue_type = None
+    if my_args.issue_type:
+        issue_type = my_args.issue_type
+        if issue_type != "pr":
+            raise ValueError("Set issue-type to 'pr' or leave empty")
+        
+
```
```suggestion
```

This will be handled directly by argparse

neubig reviewed (commented):
Hey, this is awesome, thanks so much for doing this! I left a bunch of comments below, but many of them are just part of a single section to make it so `issue_type` is not null. They can be accepted by going to files and clicking "add suggestion to batch" and then you might need to clean up some stuff I missed. I'm happy to review again when you're done revising.

Regarding the questions:

> a. Should every resolved comment open a new PR request, or update the current one?

I think update the current one. It's a little tricky because resolving the comments might mess up the current PR, but maybe that's OK since we can roll back to the previous commit?

> b. What happens when 3/5 comments are resolved? How can the user specify which code changes to push on a per comment basis?

Yeah, great question. I'd say that we might want to push anyway and notify the user that we identified 5 comments, but were only able to resolve 3. It'd be awesome if we could actually output that as a structured checklist in markdown like:

```md
I have updated the PR and resolved some of the issues that were cited in the pull request review. Specifically, I identified the following revision requests, and all the ones that I think I successfully resolved are checked off. All the unchecked ones I was not able to resolve, so manual intervention may be required:

- [ ] ...
- [x] ...
- [ ] ...
- [x] ...
- [x] ...
```

We could even provide this as an in-context example to a prompt in the `guess_success()` function that generates the PR message.

> c. How are comment dependencies handled? For instance, if two comments address the same part of the codebase, handling them separately may create merge conflicts. Do we expect the user to perform incremental commenting to mitigate this?

I think we'll want to try to resolve all comments at the same time, which will prevent this from being an issue.

malhotra5 commented:
Feedback has been addressed!

# Summary of changes
1. Prevent `issue_type` from being null
2. Follow best practice when ordering function parameters
3. Split large test cases into multiple
4. Added success indicator per review comment with explanation (initially all comments were evaluated together, but changed this to be one evaluation per comment -> see `commit #13` description for reasoning)
5. Updated  output.jsonl (`openhands_resolver/resolver_output.py`) to contain `issue-type` and success indicator per comment

# Considerations
1. I haven't implemented updating of the PR in `openhands_resolver/send_pull_request.py` after openhands resolves comments. I imagine this should be implemented in a separate PR given this one is quite long?
2. The success checklist (reference attached below) is logged inside `process_issue` in `openhands_resolver/resolve_issue.py`; should this be moved to `openhands_resolver/visualize_resolver_output.py` instead?

Success checklist reference:
```
I have updated the PR and resolved some of the issues that were cited in the pull request review. Specifically, I identified the following revision requests, and all the ones that I think I successfully resolved are checked off. All the unchecked ones I was not able to resolve, so manual intervention may be required:

- [ ] ...
- [X] ...
- [ ] ...
- [X] ...
- [X] ...

```

neubig reviewed (approved):
This looks great, thanks!

There are lots of "if" statements switching between issue and PR that we might want to refactor a little later to simplify the logic, but overall this is a good change and we should get it in.

I agree that `send_pull_request.py` could be made in a separate PR.
